### PR TITLE
Implement the FileSystem function rmTree

### PR DESCRIPTION
**Pull request by lydia-duncan**
rmTree takes in a directory and cleans up it and all of its contents.  If what
is provided is not a directory, rmTree with exit with an error message.

Add three tests of rmTree: one of removing an empty directory, one of removing
a large directory with various states for subdirectories, and one of removing
a directory which contains a symlink to another file in a separate directory.

Passed full linux testing, and the test/modules/standard/FileSystem directory for
darwin and cygwin

### CODE REVIEW

File: modules/standard/FileSystem.chpl [hunk 1]
```diff
@@ -1114,6 +1114,59 @@ proc remove(name: string) {
...
+  }
+  // Once everything else has been removed, remove ourself.
+  remove(error, root);
+}
+
+pragma "no doc"
+proc rmTree(out error: syserr, root: string) {
+  var expectedErrorCases = !exists(error, root);
+  if (error != ENOERR) then return; // Some error occurred in checking the existence of root.
+  else if (expectedErrorCases) {
```
Comment (mppf) [inline modules/standard/FileSystem.chpl]: I find the variable name "expectedErrorCases" confusing.
I'd prefer different variables like "rootMissing" or
"rootNotDir".

**Comment by mppf** [comment]
Looks good, other than my minor variable name concern. I appreciate that you thought about symbolic links (and I think handled them correctly).

**Comment by lydia-duncan** [comment]
Hmm, thinking about it more, I'm not confident I handled symlinks to directories correctly.  I'm going to write another test and get back to you on that.

**Comment by mppf** [comment]
I thought that the first loop would remove symbolic links to directories? Is that wrong? Then they would not be present for the 2nd loop that only finds directories...

**Comment by lydia-duncan** [comment]
It has to do with how listdir traverses the structure.  It only returns files in one case and directories in the other, and symlinks to directories return true from isDir, so it'd be missed in the file case.  I think.  It's an easy fix if that is the case

**Comment by lydia-duncan** [comment]
isFile and isDir use stat under the covers, and stat traverses through a link rather than examining the link itself, so it says isFile or isDir based on if what is linked to is a file or a directory.

**Comment by lydia-duncan** [comment]
Yup, my test shows that it fails to remove links to directories.  Fixing ...

**Comment by mppf** [comment]
Right, I'm pretty sure we want to remove such a link but not traverse into it and remove further things inside the linked-to directory...

**Comment by lydia-duncan** [comment]
Yup!  The fix is really easy, I'm just going to turn on links for the dir traversal and do a check on if dirname refers to a link, then switch between removing the link or traversing into the directory structure.

**Comment by lydia-duncan** [comment]
Okay @mppf, I've pushed the change in.  It passed testing locally and I'm ensuring that a checked out version of my branch works as well.

**Comment by lydia-duncan** [comment]
Testing passed :)

**Comment by mppf** [comment]
Looks good to me.

**Comment by lydia-duncan** [comment]
Awesome, thanks for the review!

**Pull request merged**